lib/memjs/memjs.js

```diff
@@ -73,19 +74,26 @@ Client.create = function(serversStr, options) {
   serversStr = serversStr || process.env.MEMCACHIER_SERVERS ||
                              process.env.MEMCACHE_SERVERS || 'localhost:11211';
   var serverUris = serversStr.split(',');
+  var serversLookup = {};
   var servers = serverUris.map(function(uri) {
     var uriParts = uri.split('@');
     var hostPort = uriParts[uriParts.length - 1].split(':');
     var userPass = (uriParts[uriParts.length - 2] || '').split(':');
+    serversLookup[uri] = server;
     return new Server(hostPort[0], parseInt(hostPort[1] || 11211, 10), userPass[0], userPass[1], options);
   });
-  return new Client(servers, options);
+  return new Client(servers, options, serversLookup);
 };
 
 // Chooses the server to talk to by hashing the given key.
 Client.prototype.server = function(key) {
   // TODO(alevy): should use consistent hashing and/or allow swapping hashing
   // mechanisms
+  // allow use of a pluggable HashRing ie: https://github.com/3rd-Eden/node-hashring
+  if (this.HashRing) {
+    const serverUri = this.HashRing.get(key);
+    return this.serversLookup[serverUri]
```

Comment:
This bypasses the server timeout and failover logic which starts 4 lines down.

You could use a server lookup that maps URI to index instead of URI to server object. Then you get `idx` conditionally from the HashRing or from the 3 lines below.

Reply:
Seems fair to me. Should I cancel this PR and submit a new one then? (Sorry I'm still a bit lost with PRs).

Reply:
Looking at it more it may be that I am confused how the `wakeUp` etc is meant to work.

With a HashRing you are hoping to get the "correct" server ie the same one that you hit last time with a specific key. If we go through another server you will get a cache miss.

Reply:
So `memjs` uses a modulo hashing scheme (I marked the code where that is implemented). A hash ring is a consistent hashing scheme. Both will always give you the same server for a particular key. The difference is what happens when a server is added or removed, i.e., when you grow or shrink your cluster of Memcached servers. A consistent hashing scheme will keep much more keys at the same server. (I wrote a post about consistent hashing a while ago which might be interesting for you: https://blog.memcachier.com/2017/09/01/maglev-our-new-consistent-hashing-scheme/).

The `wakeUp` is part of the failover implementation. When a server becomes unavailable and you have failover enabled, the server will be marked as dead and wake up after a while to be tried again. This is not really important for the pluggable hashing function though.

I just realized that my original suggestion on how to fix this PR is not ideal. Ideally, you would want a pluggable `getServer(servers, key)` function which comes with the default modulo implementation.

Reply:
> Seems fair to me. Should I cancel this PR and submit a new one then? (Sorry I'm still a bit lost with PRs).

Just push a new commit and the PR will update automatically.